devlib/collector/perfetto.py

```diff
@@ -0,0 +1,72 @@
+#    Copyright 2023 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import time
+import subprocess
+
+from devlib.collector import (CollectorBase, CollectorOutput,
+                              CollectorOutputEntry)
+from devlib.exception import TargetStableError, HostError
+from devlib.utils.misc import check_output, which, memoized
+
+OUTPUT_PERFETTO_TRACE = 'trace.perfetto-trace'
+
+class PerfettoCollector(CollectorBase):
+
+    def __init__(self, target, config=None):
+        super(PerfettoCollector, self).__init__(target)
+        self.target_output_file = target.path.join(self.target.working_directory, OUTPUT_PERFETTO_TRACE)
+        self.output_path = None
+        self.config = config
+
+        if not self.target.is_installed('perfetto'):
+            raise TargetStableError('No perfetto found on device.')
+        self.target_binary = 'perfetto'
+
+    def start(self):
+        self.start_time = time.time()
+
+        # start tracing and save the PID to stop it later
+        self.pid = self.target.execute(
+            "cat {} | {} --background --txt -c - -o {}".format(
+                self.config, self.target_binary, self.target_output_file
+            )
+        )
+
+    def stop(self):
+        # stop tracing by killing the process
+        self.target.execute("kill {}".format(self.pid))
+        self.stop_time = time.time()
+
+    def set_output(self, output_path):
+        if os.path.isdir(output_path):
+            output_path = os.path.join(output_path,
+                                       os.path.basename(self.target_output_file))
+        self.output_path = output_path
+
+    def get_data(self):
+        if self.output_path is None:
+            raise RuntimeError("Output path was not set.")
+        # adjust the pull timeout based on the size of the file
+        pull_timeout = 10 * (self.stop_time - self.start_time)
+        self.target.pull(self.target_output_file, self.output_path, timeout=pull_timeout)
```

Comment:
Do we need to provide a timeout here? If not specified devlib should fall back to using the polling mechanism which *should* be more reliable. 
Have you seen behavior contrary to this?

Reply:
That's what the FtraceCollector does in get_data() so I assumed there was a reason there and did the same for consistency

Reply:
Ah, I guess that means the FtraceCollector has not been updated since this functionality has been added. Thanks for pointing that out, we should probably get that updated as well.

Reply:
Yep, I think that would be good thanks.

Reply:
@mrkajetanp Do you have any logs of the runs where timeouts were occuring to take a look at?